### PR TITLE
DAOS-3676 Test: Verify container destroy during aggregation

### PR DESCRIPTION
**Pull request by ravalsam**
Adding Yield in aggrgeation code path so the container can be destroyed
while aggregation is in progress.

Signed-off-by: Samir Raval <samir.raval@intel.com>

### CODE REVIEW

File: src/tests/suite/daos_container.c [hunk 1]
```diff
@@ -2203,6 +2204,52 @@ co_rf_simple(void **state)
...
+			      DAOS_CONT_AGG_YIED | DAOS_FAIL_ALWAYS, 0, NULL);
+
+	/* Write/fetch and Punch Data with 2K size */
+	for (i = 0; i <= 50; i++)
+		io_simple_internal(state, oid, IO_SIZE_SCM * 32, DAOS_IOD_ARRAY,
+				   "io_simple_scm_array dkey",
+				   "io_simple_scm_array akey");
+
+	/**
+	 * Run Pool query every 5 seconds for Total 40 seconds
```
Comment (NiuYawei) [inline src/tests/suite/daos_container.c:2233]: Do we really need to wait that long? 25 seconds isn't enough?

File: src/vos/vos_aggregate.c [hunk 2]
```diff
@@ -1749,6 +1749,10 @@ vos_agg_ev(daos_handle_t ih, vos_iter_entry_t *entry,
 		return rc;
 	}
 
+	/* Aggregation Yield for testing purpose */
+	while (DAOS_FAIL_CHECK(DAOS_CONT_AGG_YIED))
+		ABT_thread_yield();
```
Comment (NiuYawei) [inline src/vos/vos_aggregate.c:1754]: It's better mark VOS_ITER_CB_YIELD by "*act |= VOS_ITER_CB_YIELD". (though your test doesn't update VOS tree while yielding)

File: src/tests/suite/daos_container.c [hunk 3]
```diff
@@ -2203,6 +2204,52 @@ co_rf_simple(void **state)
...
+delet_container_during_aggregation(void **state)
+{
+	test_arg_t	*arg = *state;
+	daos_obj_id_t	 oid;
+	daos_pool_info_t pinfo;
+	int		i;
+	int		rc;
+
+	/* Prepare records */
+	oid = dts_oid_gen(OC_RP_2G1, 0, arg->myrank);
```
Comment (NiuYawei) [inline src/tests/suite/daos_container.c:2217]: That implies the test can only run on nodes >= 2, maybe we can use OC_SX that the test can run on single node as well.

File: src/tests/suite/daos_container.c [hunk 4]
```diff
@@ -2203,6 +2204,52 @@ co_rf_simple(void **state)
...
+
+	print_message("Initial Pool Query\n");
+	pool_storage_info(state, &pinfo);
+
+	/* Aggregation will be Hold */
+	daos_debug_set_params(arg->group, -1, DMG_KEY_FAIL_LOC,
+			      DAOS_CONT_AGG_YIED | DAOS_FAIL_ALWAYS, 0, NULL);
+
+	/* Write/fetch and Punch Data with 2K size */
+	for (i = 0; i <= 50; i++)
```
Comment (NiuYawei) [inline src/tests/suite/daos_container.c:2227]: Maybe larger dataset to make aggregation last longer?

File: src/include/daos/common.h [hunk 5]
```diff
@@ -677,6 +677,7 @@ enum {
 #define DAOS_CONT_CLOSE_FAIL_CORPC	(DAOS_FAIL_UNIT_TEST_GROUP_LOC | 0x66)
 #define DAOS_CONT_QUERY_FAIL_CORPC	(DAOS_FAIL_UNIT_TEST_GROUP_LOC | 0x67)
 #define DAOS_CONT_OPEN_FAIL		(DAOS_FAIL_UNIT_TEST_GROUP_LOC | 0x68)
+#define DAOS_CONT_AGG_YIED		(DAOS_FAIL_UNIT_TEST_GROUP_LOC | 0x69)
```
[the remark this answers is not in the document]
Comment (ravalsam) [reply src/include/daos/common.h:680]: sure will combine.

File: src/tests/suite/daos_container.c (hunk 3 above)
Comment (ravalsam) [reply src/tests/suite/daos_container.c:2217]: oh sure.

File: src/tests/suite/daos_container.c (hunk 4 above)
Comment (ravalsam) [reply src/tests/suite/daos_container.c:2227]: I tested and this worked so though not to write more data but can increase

File: src/tests/suite/daos_container.c (hunk 1 above)
Comment (ravalsam) [reply src/tests/suite/daos_container.c:2233]: Yes, I wanted to run top of aggregation time to be sure it's about to start. And is it 25 or 30 seconds?

File: src/vos/vos_aggregate.c (hunk 2 above)
Comment (ravalsam) [reply src/vos/vos_aggregate.c:1754]: hmm ok, let me understand.

File: src/tests/suite/daos_container.c (hunk 4 above)
Comment (NiuYawei) [reply src/tests/suite/daos_container.c:2227]: The test is like: 1. block aggregation; 2. unblock aggregation; 3. destroy container; so I hope the aggregation not finishing before the 3rd step. Maybe larger dataset could be helpful.

File: src/tests/suite/daos_container.c (hunk 1 above)
Comment (NiuYawei) [reply src/tests/suite/daos_container.c:2233]: I think 30 seconds should be enough.

File: src/vos/vos_aggregate.c (hunk 2 above)
Comment (NiuYawei) [reply src/vos/vos_aggregate.c:1754]: I mean adding a "*act |= VOS_ITER_CB_YIELD" after this ABT_thread_yield().

**Review by jolivier23 — APPROVED** [review]

**Comment by ravalsam** [comment]
There is a single test failure and looks similar to the existing defect DAOS-6367.

**Comment by kprantis** [comment]
Is this patch now just waiting for reviews?

**Comment by ravalsam** [comment]
> Is this patch now just waiting for reviews?

Yes, Niu has reviewed it initially but he is out on Holiday, so once he is back will get it reviewed.

**Review by NiuYawei — DISMISSED** [review]

File: src/tests/suite/daos_container.c [hunk 6]
```diff
@@ -2203,6 +2204,52 @@ co_rf_simple(void **state)
...
+		pool_storage_info(state, &pinfo);
+		sleep(5);
+	}
+
+	/* Aggregation will continue */
+	daos_debug_set_params(arg->group, -1, DMG_KEY_FAIL_LOC, 0, 0, NULL);
+
+	/* Destroy the container while Aggregation is running */
+	rc = test_teardown_cont(arg);
+	assert_int_equal(rc, 0);
```
Comment (jolivier23) [inline src/tests/suite/daos_container.c:2246]: assert_rc_equal

**Review by jolivier23 — CHANGES_REQUESTED** [review]
I'd like to ensure this doesn't create instability with that

**Review by jolivier23 — APPROVED** [review]

**Review by NiuYawei — APPROVED** [review]

**Pull request merged**